sui-execution/latest/sui-move-natives/src/crypto/sui_state_proof.rs

```diff
@@ -0,0 +1,387 @@
+        ));
+    };
+
+    let Ok(summary) = bcs::from_bytes::<CertifiedCheckpointSummary>(&summary_bytes) else {
+        return Ok(NativeResult::err(
+            cost,
+            INVALID_CHECKPOINT_SUMMARY
+        ));
+    };
+
+    let Ok(checkpoint_contents) = bcs::from_bytes::<CheckpointContents>(&checkpoint_contents_bytes) else {
+        return Ok(NativeResult::err(
+            cost,
+            INVALID_INPUT
+        ));
+    };
+
+    let Ok(transaction) = bcs::from_bytes::<CheckpointTransaction>(&transaction_bytes) else {
+        return Ok(NativeResult::err(
+            cost,
+            INVALID_INPUT
+        ));
+    };
+
+
+    let Ok(type_layout) = bcs::from_bytes::<MoveTypeLayout>(&type_layout_bytes) else {
+        return Ok(NativeResult::err(
+            cost,
+            INVALID_INPUT
+        ));
+    };
+
+
+    let Ok(package_id_target) = bcs::from_bytes::<ObjectID>(&package_id_bytes) else {
+        return Ok(NativeResult::err(
+            cost,
+            INVALID_INPUT
+        ));
+    };
+    
+    // Verify the checkpoint summary using the committee
+    let res = summary.verify_with_contents(&committee, Some(&checkpoint_contents));
+    if let Err(_) = res {
+        return Ok(NativeResult::err(cost, INVALID_TX));
+    }
+
+
+    // Ensure the tx is part of the checkpoint
+    let is_valid_checkpoint_tx = checkpoint_contents.iter().any(|&digest| digest == transaction.effects.execution_digests());
+    if !is_valid_checkpoint_tx {
+      return Ok(NativeResult::err(cost, INVALID_TX));
+    };
+
+    let tx_events = &transaction.events.as_ref().unwrap().data;
+
+    let results: Vec<(SuiAddress, Vec<u8>)> = tx_events.into_iter().filter_map(|event| {
+        if event.package_id == package_id_target && event.type_.module.clone().into_string() == "dwallet_cap" && event.type_.name.clone().into_string() == "DWalletNetworkApproveRequest" {
+            let json_val = SuiJsonValue::from_bcs_bytes(Some(&type_layout), &event.contents).unwrap().to_json_value();
+    
+            let cap_id_str = json_val.get("cap_id").and_then(JsonValue::as_str);
+            let cap_id = cap_id_str.and_then(|s| SuiAddress::from_str(s).ok());
+    
+            let approve_message = json_val.get("message").and_then(JsonValue::as_array);
+            let approve_msg_vec: Option<Vec<u8>> = approve_message.map(|msg_array| {
+                msg_array.iter().map(|msg| {
+                    msg.as_u64().unwrap() as u8
+                }).collect()
+            });
+    
+            if let (Some(cap_id), Some(approve_msg_vec)) = (cap_id, approve_msg_vec) {
+                Some((cap_id, approve_msg_vec))
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }).collect();
+    
+    let (cap_ids, messages): (Vec<_>, Vec<_>) = results.into_iter().unzip();
+    
+    if cap_ids.len() != messages.len() {
+        return Ok(NativeResult::err(cost, INVALID_TX));
+    }
```

Comment:
This check is redundant since you already construct it in a way that assures this condition will always be fulfilled (as pairs -> unwrap). I don't see a problem with leaving this sanity check anyways though.